### PR TITLE
only selectively kill child, not entire process group

### PR DESCRIPTION
**Pull request by kroening**
This is a new approach to addressing the problem explained in PR #319.

It is enabled by the fact that SMT solvers are now executed with run(), as opposed to system().

### CODE REVIEW

File: src/util/run.cpp [hunk 1]
```diff
@@ -202,6 +207,8 @@ int run(
           return 1;
         }
 
+      unregister_child();
```
Comment (tautschnig) [inline src/util/run.cpp]: Asking for an unrelated fix: could the `while` loop above please get braces around its body? The jump in indentation looked pretty awkward at first sight.

File: src/util/signal_catcher.cpp [hunk 2]
```diff
@@ -21,11 +21,21 @@ Author: Daniel Kroening, kroening@kroening.com
 
 // Here we have an instance of an ugly global object.
 // It keeps track of any child processes that we'll kill
-// when we are told to terminate.
+// when we are told to terminate. "No child" is indicated by '0'.
 
 #ifdef _WIN32
 #else
-std::vector<pid_t> pids_of_children;
+pid_t child_pid = 0;
```
Comment (tautschnig) [inline src/util/signal_catcher.cpp:25]: `#include <vector>` is probably redundant now. Actually more of the includes could be cleaned up (we don't need anything on Windows, `csignal` is included by the header file already).

**Review by tautschnig — APPROVED** [review]
So ... does it actually work as expected? One experiment is invoking CBMC from Java, then sending a TERM signal to the CBMC process -> the Java process should still terminate ok and not be killed.

**Comment by tautschnig** [comment]
> It is enabled by the fact that SMT solvers are now executed with run(), as opposed to system().

Note that this is only part of the story: we still use `system` in places, as well as `popen`. All of those should be changed to `run` as well.

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: c93180ac7e81d12f3a37279635c2c646c21536f6).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/84146654

File: src/util/signal_catcher.cpp [hunk 3]
```diff
@@ -11,21 +11,27 @@ Author: Daniel Kroening, kroening@kroening.com
...
+
+void register_child(pid_t pid)
+{
```
Comment (krobelus) [inline src/util/signal_catcher.cpp:28]: We might want to include an invariant ensuring that only one child is registered.
If multiple child processes are actually supported here, then some kind of warning would be useful.

Comment (kroening) [reply src/util/signal_catcher.cpp:28]: added preconditions

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 1c8fc56010dfca620a5357c62013380c095ec78c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/84692107
